[UR][CI] Add first version of UR workflow

lukaszstolarczuk opened the pull request:
~~Last commit included just to verify if `test_job` will be executed.~~

With extra dummy changes (now missing in PR) we just didn't skip the `test_job`, e.g. here: https://github.com/intel/llvm/actions/runs/13040807649/job/36382037543

kbenzie commented on .github/workflows/sycl-linux-precommit.yml:
```diff
@@ -177,6 +177,14 @@ jobs:
       sycl_toolchain_archive: ${{ needs.build.outputs.artifact_archive_name }}
       sycl_toolchain_decompress_command: ${{ needs.build.outputs.artifact_decompress_command }}
 
+  # First approach to UR workflow - it's executed on any sycl changes;
+  # It downloads the UR sources from the original repo and builds it
+  # i.a. on UR hardware.
+  unified-runtime:
+    name: Unified Runtime
+    needs: [build, detect_changes]
+    if: ${{ always() && !cancelled() && contains(needs.detect_changes.outputs.filters, 'sycl') }}
+    uses: ./.github/workflows/unified-runtime.yml
```
I think we should keep this separate from the *SYCL Pre Commit on Linux* workflow for the first iteration. When looking at the [checks](https://github.com/intel/llvm/pull/16827/checks) tab we should have a separate *Unified Rutime Pre Commit* section on the left.

Further to this, one of the suggestions made to us is to only run the UR build/test workflows when the UR source has changed. This will probably require the code to have actually moved into intel/llvm first in order to actually trigger job creation.

lukaszstolarczuk replied:
make sense, updated.

as for detecting UR changes - I updated "detect_changes" workflow, and added a dummy UR changes to check trigger (it's the last commit).

kbenzie commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,125 @@
+    runs-on: ${{inputs.runner_name}}
+
+    steps:
+    # TODO: If UR is merged into llvm it will require changes:
+    # - checkout llvm instead of UR
+    # - `cd sycl` and configure UR project from local tree
+    #
+    # Also, the step of downloading DPC++ should be integrated somehow;
+    # either compile it or use cached binaries (if accessible)...?
+    #
+    - name: Checkout UR
+      uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+      with:
+        repository: 'oneapi-src/unified-runtime'
+        ref: main
```
Once the move has been made the UR source will exits in the `unified-runtime` subdirectory of the intel/llvm checkout, can we specify the checkout to be placed there to validate the rest of the workflow?

lukaszstolarczuk replied:
I updated the checkout place via `path: unified-runtime`, not sure if this is change you had in mind.

kbenzie replied:
Yes, this makes sense. Once we pull in the UR source this should be the same path so the rest of the workflow shouldn't need to change.

kbenzie commented on sycl/cmake/modules/UnifiedRuntimeTag.cmake:
```diff
@@ -1,7 +1,6 @@
+# Date:   Wed Jan 22 11:06:11 2025 +0100
+# [common] Bump UMF to early 0.11 version, from main
+# It includes i.a. MacOS fix for compiler.
+set(UNIFIED_RUNTIME_TAG d193046de592482c47d87fdfaf92c7b8c59c9b66)
```
Is this tag update necessary for the workflow changes?

lukaszstolarczuk replied:
just a dummy change to trigger `sycl` changes (in detect_changes job) - replaced now with dummy changes for `unified-runtime` dir - this commit will be removed before merging, of course

kbenzie commented on .github/workflows/unified-runtime.yml:
```diff
@@ -0,0 +1,127 @@
+# It was pretty much copy-pasted from UR repository.
+# Over time it will be split and merged into existing sycl workflows.
+
+on: workflow_call
```
As mentioned above, I think we should define separate triggers here rather than making this called by the *SYCL Pre Commit on Linux* workflow.

kbenzie commented:
During the meeting we think that we should merge this before the PR which pulls in the UR source code once we have the jobs working correctly. Then, in the time between merging this and pulling in the UR sources we disable the workflow to avoid overloading the runners.

lukaszstolarczuk commented:
So, FYI, the current status is that the CUDA runner went wild, not sure what happened (I'm betting a bad `apt` update). Working on reviving the machine, if nothing succeed I'll replace it with a different runner...

lukaszstolarczuk commented:
there's new CUDA runner enabled, we should be good now.

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+name: UR - Build adapters on HW
```
maybe a nit, but what does it mean to build an adapter on hardware?

kbenzie replied:
In this case it means run the tests on hardware, these jobs don't currently have a build/test separation. The name can be updated:

```suggestion
name: UR - Build adapters, test on HW
```

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+
+    - name: Download DPC++
+      run: |
+        wget -O ${{github.workspace}}/dpcpp_compiler.tar.gz https://github.com/intel/llvm/releases/download/nightly-2024-12-12/sycl_linux.tar.gz
```
do we want to use the same dpcpp version every time or do we want to use the latest nightly

also we could use a docker image that has it preinstalled

kbenzie replied:
Moving forwards we should definitely use the nightly. We had issues in the past which is why these jobs use a fixed version for now.

I think, for the same of ensuring as smooth a transition as possible, we should leave this fixed version and work on use the nightly post merge.

The work on enable use of docker images on these runners is ongoing.

aelovikov-intel commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+name: UR - Build adapters on HW
+
+on:
+  workflow_call:
```
We prefer helper workflows like this one to have an additional workflow_dispatch trigger for testing/experimenting, e.g. https://github.com/intel/llvm/blob/822cf9b028726b730bb4156a0704acd8201f69aa/.github/workflows/sycl-linux-run-tests.yml#L106

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+        -DUR_DPCXX=${{github.workspace}}/dpcpp_compiler/bin/clang++
+        -DUR_SYCL_LIBRARY_DIR=${{github.workspace}}/dpcpp_compiler/lib
+        -DCMAKE_INSTALL_PREFIX=${{github.workspace}}/install
+        ${{ matrix.adapter.name == 'HIP' && '-DUR_CONFORMANCE_AMD_ARCH=gfx1030' || '' }}
```
im not sure i understand why we need to set `DUR_CONFORMANCE_AMD_ARCH` here

kbenzie replied:
This might be a hang over, I think we autodetect the AMD GPU arch for compiling test programs for the device now.

kbenzie replied:
I think we should keep this for now and investigate if we can remove it post merge, can you add this to the TODO list @lukaszstolarczuk ?

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+
+    - name: Build
+      # This is so that device binaries can find the sycl runtime library
+      run: cmake --build ${{github.workspace}}/build -j $(nproc)
```
if you use ninja instead of make as the cmake generated at the configure step (`-GNinja`) you shouldnt need this, and i think ninja is just better anyway

kbenzie replied:
Seems like the idea is to replace the Unix Makefiles jobs with Ninja jobs.

aelovikov-intel commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+    # - checkout only llvm repo
+    # - configure UR project from local tree
+    #
+    # Also, the step of downloading DPC++ should be integrated somehow;
+    # either compile it or use cached binaries (if accessible)...?
```
Nightly "release" maybe? We don't try re-use post-commit binaries (although artifacts might actually be available).

kbenzie replied:
Nightly would make sense, yeah. A related comment [here](https://github.com/intel/llvm/pull/16827/files#r1954752515).

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+
+    - name: Test adapter specific
+      working-directory: ${{github.workspace}}/build
+      run: ctest -C ${{matrix.build_type}} --output-on-failure -L "adapter-specific" -E "memcheck" --timeout 600
```
do we have to set the wd or can we just specify the path when running the tests?

kbenzie replied:
Looks like we could add the `--test-dir ${{github.workspace}}/build` option.

We do intend to replae ctest with lit after the repo move so this will be temporary.

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+    - name: Get information about platform
+      working-directory: ${{github.workspace}}/unified-runtime
+      if: ${{ always() }}
+      run: .github/scripts/get_system_info.sh
```
is this just for debugging output?

kbenzie replied:
It has been very helpful for diagnosing job failures, yes.

sarnex replied:
if you think it will be useful for end users, like if their ci fails to help them debug it or somehing, then imo we can leave it

sarnex commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+    - name: Get information about platform
+      working-directory: ${{github.workspace}}/unified-runtime
+      if: ${{ always() }}
+      run: .github/scripts/get_system_info.sh
```
i dont think this file exists

kbenzie replied:
This is in the UR repo https://github.com/oneapi-src/unified-runtime/blob/main/.github/scripts/get_system_info.sh, its found because of the `working-directory` line pointing to the UR clone which exists in the same location as it will post repo move.

aelovikov-intel commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+
+    - name: Configure CMake
+      working-directory: ${{github.workspace}}/unified-runtime
+      run: >
```
```suggestion
      run: |
```

or a comment why different from other steps/workflows.

kbenzie commented:
Failures are known flaky:

* https://github.com/intel/llvm/issues/16978
* https://github.com/intel/llvm/issues/16951

sarnex commented on .github/workflows/ur-precommit.yml:
```diff
@@ -0,0 +1,193 @@
+  # - test_job should be removed
+  #
+  test_job:
+    name: UR test job
```
should we instead say `Detect UR changes`?

aelovikov-intel commented on .github/workflows/ur-build-hw.yml:
```diff
@@ -0,0 +1,132 @@
+
+    - name: Build
+      # This is so that device binaries can find the sycl runtime library
+      run: cmake --build ${{github.workspace}}/build -j $(nproc)
```
Can we just use `ninja` instead of `make` as generator and then
```suggestion
      run: cmake --build ${{github.workspace}}/build
```

sarnex commented on .github/workflows/ur-precommit.yml:
```diff
@@ -0,0 +1,193 @@
+      adapter_name: L0_V2
+      runner_name: UR_L0
+
+  level_zero_static:
```
we could use a matrix for all of the adapters

aelovikov-intel replied:
```
    strategy:
      fail-fast: false
      matrix:
        include:
```
? That's what we use in E2E tasks.

sarnex replied:
right, we do the same thing with a matrix

kbenzie commented:
@intel/dpcpp-devops-reviewers please review

sarnex commented on .github/workflows/ur-source-checks.yml:
```diff
@@ -0,0 +1,83 @@
+        Expand-Archive -Path "$WorkingDir\doxygen.zip"
+        Add-Content $env:GITHUB_PATH "$WorkingDir\doxygen"
+
+    - name: "[Lin] Install hwloc"
```
imo we should split the win/lin steps into seperate workflows

aelovikov-intel replied:
I'm willing to have them in a single file for now. The benefits for SYCL were:
1) Ability to cancel individual Linux/Windows if some CI job isn't necessary
2) Job dispatch control via "Environment" for when we had zombie processes post testing making the runner unusable for subsequent tasks.

If UR CI won't be affected by these (I'd assume it won't), then there will be little value in splitting the OSes.

sarnex replied:
yeah my comment was mainly about confusion reading the results because people are used to how it's done for sycl, but its not a big deal

lukaszstolarczuk replied:
added to the TODO list

sarnex reviewed (commented):
first pass!

aelovikov-intel commented on .github/workflows/ur-precommit.yml:
```diff
@@ -0,0 +1,193 @@
+    needs: [detect_changes]
+    uses: ./.github/workflows/ur-source-checks.yml
+
+  level_zero:
+    name: Level Zero
+    needs: [source_checks]
+    uses: ./.github/workflows/ur-build-hw.yml
+    with:
+      adapter_name: L0
+      runner_name: UR_L0
+
+  level_zero_v2:
+    name: Level Zero V2
+    needs: [source_checks]
+    uses: ./.github/workflows/ur-build-hw.yml
+    with:
+      adapter_name: L0_V2
+      runner_name: UR_L0
+
+  level_zero_static:
+    name: Level Zero static
+    needs: [source_checks]
+    uses: ./.github/workflows/ur-build-hw.yml
+    with:
+      adapter_name: L0
+      runner_name: UR_L0
+      static_loader: ON
+      static_adapter: ON
+
+  opencl:
+    name: OpenCL
+    needs: [source_checks]
+    uses: ./.github/workflows/ur-build-hw.yml
+    with:
+      adapter_name: OPENCL
+      runner_name: UR_OPENCL
+      platform: "Intel(R) OpenCL"
```
Why isn't this a `matrix`?

sarnex commented on .github/workflows/ur-precommit.yml:
```diff
@@ -0,0 +1,141 @@
+          {name: L0, runner: UR_L0},
+          {name: L0_V2, runner: UR_L0},
+          {name: L0, runner: UR_L0, static: ON},
+          {name: OPENCL, runner: UR_OPENCL, platform: "Intel(R) OpenCL"},
```
setting the platform name with a string seems error prone, if the spacing or anything is wrong it will fail right? maybe we could use a single word like `opencl` that, if required, gets updated to the required value for the actual build by the ur-build workflow.

kbenzie commented:
Somethings gone wrong with the HIP job https://github.com/intel/llvm/actions/runs/13313402495/job/37182358932?pr=16827

kbenzie commented:
@sarnex @aelovikov-intel is it correct there's a public holiday in the US on Monday? If so, I would appriciate an approval so that we can merge this once we've address the outstanding issues ahead of the repo move on Tuesday.

kbenzie commented:
> yeah theres a holiday on mon, discussed it with andrei offline, but please address all feedback in a follow up

Yes for sure, thank you!

sarnex commented:
@kbenzie sorry to be clear, is it okay if i merge this now? we would prefer that to catch any issues while we are working. unaddressed feedback is expected to be addressed in followup pr

sarnex commented:
oh i see the UR hip is failing, probably that needs to be fixed first

kbenzie commented:
> @kbenzie sorry to be clear, is it okay if i merge this now? we would prefer that to catch any issues while we are working. unaddressed feedback is expected to be addressed in followup pr

There's an issue with the HIP runner which @lukaszstolarczuk is looking into.

sarnex commented:
ok cool just ping me when its ready for merge today, or ask a gatekeeper not in US to merge on monday

lukaszstolarczuk commented:
> @kbenzie sorry to be clear, is it okay if i merge this now? we would prefer that to catch any issues while we are working. unaddressed feedback is expected to be addressed in followup pr

I already have some changes implemented locally. For me it's no difference, I can open new PR with changes.
Probably would be better to speed up the process and not block Tuesday's move - so we can merge today.

kbenzie commented:
If the HIP runner issue isn't resolved until Monday I'll be sure to keep a close eye on things post merge in case something goes awry. Otherwise if you fix it today @lukaszstolarczuk please go ahead with the request to merge.

I need to log off for now, I'll be back online briefly in a few hours to check on things.

kbenzie commented:
> > @kbenzie sorry to be clear, is it okay if i merge this now? we would prefer that to catch any issues while we are working. unaddressed feedback is expected to be addressed in followup pr
> 
> I already have some changes implemented locally. For me it's no difference, I can open new PR with changes. Probably would be better to speed up the process and not block Tuesday's move - so we can merge today.

Okay, lets keep it separate to derisk the move.

lukaszstolarczuk commented:
> If the HIP runner issue isn't resolved until Monday I'll be sure to keep a close eye on things post merge in case something goes awry. Otherwise if you fix it today @lukaszstolarczuk please go ahead with the request to merge.
> 
> I need to log off for now, I'll be back online briefly in a few hours to check on things.

I think I fixed the HIP issue - could you please re-run the HIP job?

kbenzie commented:
Done, I actually need to go now. Thank you!

lukaszstolarczuk reviewed (commented):
I only pushed for now the `fail-fast: false` and TODO comments. I'm good with following up in the next PR, next week.

@sarnex, If you're good with the few red E2E tests on CI, please merge this.

sarnex commented:
@lukaszstolarczuk So i have a PR [here](https://github.com/intel/llvm/pull/16972), and it looks like UR CI is [running](https://github.com/intel/llvm/actions/runs/13335949937/job/37251086700?pr=16972) even though it wasn't changed. Is that expected?

kbenzie commented:
I've disabled the workflow for now, good to see it working. I'll reneable it before we the repo move.
